Document the SessionDiagnostic derive macro. (#961)
rust-lang/rustc-dev-guide: src/diagnostics/sessiondiagnostic.md

The SessionDiagnostic derive macro gives an alternate way to the DiagnosticBuilder API for defining
and emitting errors.  It allows a struct to be annotated with information which allows it to be
transformed and emitted as a Diagnostic.

As an example, we'll take a look at how the "field already declared" diagnostic is actually defined
in the compiler (see the definition
[here](https://github.com/rust-lang/rust/blob/75042566d1c90d912f22e4db43b6d3af98447986/compiler/rustc_typeck/src/errors.rs#L65-L74)
and usage
[here](https://github.com/rust-lang/rust/blob/75042566d1c90d912f22e4db43b6d3af98447986/compiler/rustc_typeck/src/collect.rs#L863-L867)):

```rust,ignore
#[derive(SessionDiagnostic)]
#[error = "E0124"]
pub struct FieldAlreadyDeclared {
    pub field_name: Ident,
    #[message = "field `{field_name}` is already declared"]
    #[label = "field already declared"]
    pub span: Span,
    #[label = "`{field_name}` first declared here"]
    pub prev_span: Span,
}
// ...
tcx.sess.emit_err(FieldAlreadyDeclared {
    field_name: f.ident,
    span: f.span,
    prev_span,
});
```

We see that using `SessionDiagnostic` is relatively straight forward.  The `#[error = "..."]`

> attribute is used to supply the error code for the diagnostic.  We are then annotate fields in the

Reviewer: ```suggestion
attribute is used to supply the error code for the diagnostic.  We then annotate fields in the
```

> struct with various information of how to convert an instance of the struct into a proper

Reviewer: Maybe this?

```suggestion
struct with various information on how to convert an instance of the struct into a rendered
```

diagnostic.  The attributes above produce code which is roughly equivalent to the following (in
pseudo-Rust):

```rust,ignore
impl SessionDiagnostic for FieldAlreadyDeclared {
    fn into_diagnostic(self, sess: &'_ rustc_session::Session) -> DiagnosticBuilder<'_> {
        let mut diag = sess.struct_err_with_code("", rustc_errors::DiagnosticId::Error("E0124"));
        diag.set_span(self.span);
        diag.set_primary_message(format!("field `{field_name}` is already declared", field_name = self.field_name));
        diag.span_label(self.span, "field already declared");
        diag.span_label(self.prev_span, format!("`{field_name}` first declared here", field_name = self.field_name));
        diag
    }
}
```

The generated code draws attention to a number of features.  First, we see that within the strings

> passed to each attribute, we see that field names can be referenced without needing to be passed

Reviewer: You already said “we see that”:

```suggestion
passed to each attribute, field names can be referenced without needing to be passed
```
